File: sdk_v2/js/test/detail/modelLoadManager.test.ts
import { expect } from 'chai';
import { ModelLoadManager } from '../../src/detail/modelLoadManager.js';
import { getTestManager, TEST_MODEL_ALIAS, IS_RUNNING_IN_CI } from '../testUtils.js';

describe('ModelLoadManager', function() {
    let coreInterop: any;
    let modelId: string;
    let managerInstance: any;
    let serviceUrl: string;

    before(async function() {
        managerInstance = getTestManager();
        // Access private coreInterop using any cast
        coreInterop = (managerInstance as any).coreInterop;
        
        const catalog = managerInstance.catalog;
        const model = await catalog.getModel(TEST_MODEL_ALIAS);
        if (!model) {
            throw new Error(`Model ${TEST_MODEL_ALIAS} not found in catalog`);
        }
        modelId = model.id;

        // Start the real web service if not in CI
        if (!IS_RUNNING_IN_CI) {
            try {
                managerInstance.startWebService();
                const urls = managerInstance.urls;
                if (!urls || urls.length === 0) {
                    console.warn("Web service started but no URLs returned");
                } else {
                    serviceUrl = urls[0];
                }
            } catch (e: any) {
                console.warn(`Skipping web service tests: Failed to start web service (${e.message})`);
                // If start_web_service is not supported by the local core binary, we can't run these tests.
            }
        }
    });

    after(function() {
        if (!IS_RUNNING_IN_CI && managerInstance) {
            try {
                managerInstance.stopWebService();
            } catch (e) {
                console.warn("Failed to stop web service:", e);
            }
        }
    });

    it('should load model using core interop when no external url is provided', async function() {
        this.timeout(30000);
        const manager = new ModelLoadManager(coreInterop);
        
        await manager.load(modelId);
        
        const loaded = await manager.listLoaded();
        expect(loaded).to.include(modelId);
    });

    it('should unload model using core interop when no external url is provided', async function() {
        this.timeout(30000);
        const manager = new ModelLoadManager(coreInterop);
        
        await manager.load(modelId);
        let loaded = await manager.listLoaded();
        expect(loaded).to.include(modelId);

        await manager.unload(modelId);
        
        loaded = await manager.listLoaded();
        expect(loaded).to.not.include(modelId);
    });

    it('should list loaded models using core interop when no external url is provided', async function() {
        this.timeout(30000);
        const manager = new ModelLoadManager(coreInterop);
        
        await manager.load(modelId);
        
        const loaded = await manager.listLoaded();
        expect(loaded).to.be.an('array');
        expect(loaded).to.include(modelId);
    });

    it('should load and unload model using external service when url is provided', async function() {
        if (IS_RUNNING_IN_CI || !serviceUrl) {
            this.skip();
        }
        
        const manager = new ModelLoadManager(coreInterop, serviceUrl);
        
        // Load it first so we can unload it (can use core interop to setup state)
        // Creating a manager WITHOUT serviceUrl to force core interop usage for setup
        const setupManager = new ModelLoadManager(coreInterop);
        await setupManager.load(modelId); 
        
        let loaded = await setupManager.listLoaded();
        expect(loaded).to.include(modelId);

        // Unload via external service
        await manager.unload(modelId);
        
        // Verify via core interop
        loaded = await setupManager.listLoaded();
        expect(loaded).to.not.include(modelId);
    });

    it('should list loaded models using external service when url is provided', async function() {
        if (IS_RUNNING_IN_CI || !serviceUrl) {
            this.skip();
        }
        
        const manager = new ModelLoadManager(coreInterop, serviceUrl);
        const setupManager = new ModelLoadManager(coreInterop);
        
        // Setup: Load model via core
        await setupManager.load(modelId);

        // Verify: List via external service
        const loaded = await manager.listLoaded();
        expect(loaded).to.be.an('array');
        expect(loaded).to.include(modelId);
    });
});

File: sdk_v2/rust/tests/model_load_manager_test.rs
//! Integration tests for model loading and unloading through the public API.
//!
//! Mirrors `modelLoadManager.test.ts` from the JavaScript SDK.
//!
//! **Note:** In the JavaScript SDK these tests access the private
//! `coreInterop` property via an `as any` cast.  In Rust, `CoreInterop` and
//! `ModelLoadManager::new` are `pub(crate)` and cannot be reached from
//! integration tests.  Instead, we exercise model loading and unloading
//! through the public [`Model`] and [`Catalog`] APIs which internally
//! delegate to `ModelLoadManager`.

mod common;

mod tests {
    use super::*;

    // ── Helpers ──────────────────────────────────────────────────────────

    /// Return the test model from the catalog.
    async fn get_test_model() -> foundry_local_sdk::Model {
        let manager = common::get_test_manager();
        let catalog = manager.catalog();
        catalog
            .get_model(common::TEST_MODEL_ALIAS)
            .await
            .expect("get_model failed")
    }

    // ── Core-interop path ────────────────────────────────────────────────

    /// Loading a model via the core interop (in-process) path should succeed.
    ///
    /// Timeout note: the JS test uses a 120 s timeout.
    #[tokio::test]
    #[ignore = "requires native Foundry Local library"]
    async fn should_load_model_using_core_interop() {
        let model = get_test_model().await;

        model
            .load()
            .await
            .expect("model.load() failed");
    }

    /// Unloading a previously loaded model via the core interop path should
    /// succeed.
    ///
    /// Timeout note: the JS test uses a 120 s timeout.
    #[tokio::test]
    #[ignore = "requires native Foundry Local library"]
    async fn should_unload_model_using_core_interop() {
        let model = get_test_model().await;

        // Ensure the model is loaded first.
        model
            .load()
            .await
            .expect("model.load() failed");

        model
            .unload()
            .await
            .expect("model.unload() failed");
    }

    /// Listing loaded models via the core interop path should return a
    /// collection (possibly empty, but the call itself must succeed).
    #[tokio::test]
    #[ignore = "requires native Foundry Local library"]
    async fn should_list_loaded_models_using_core_interop() {
        let manager = common::get_test_manager();
        let catalog = manager.catalog();

        let loaded = catalog
            .get_loaded_models()
            .await
            .expect("catalog.get_loaded_models() failed");

        // The result should be a valid (possibly empty) list of model IDs.
        // (Vec<String> is always valid; just ensure the call succeeded.)
        let _ = loaded;
    }

    // ── External web-service path ────────────────────────────────────────

    /// Loading and unloading a model through the external HTTP service should
    /// succeed.
    ///
    /// This test is skipped in CI because it requires a running web service.
    ///
    /// Timeout note: the JS test uses a 120 s timeout.
    #[tokio::test]
    #[ignore = "requires native Foundry Local library and running web service"]
    async fn should_load_and_unload_model_using_external_service() {
        if common::is_running_in_ci() {
            eprintln!("Skipping external-service test in CI");
            return;
        }

        let manager = common::get_test_manager();
        let model = get_test_model().await;

        // Start the web service so we can test the HTTP path.
        let _urls = manager
            .start_web_service()
            .await
            .expect("start_web_service failed");

        // Load via the model API (delegates to ModelLoadManager internally).
        model
            .load()
            .await
            .expect("load via external service failed");

        // Unload
        model
            .unload()
            .await
            .expect("unload via external service failed");
    }

    /// Listing loaded models through the external HTTP service should succeed.
    ///
    /// This test is skipped in CI because it requires a running web service.
    #[tokio::test]
    #[ignore = "requires native Foundry Local library and running web service"]
    async fn should_list_loaded_models_using_external_service() {
        if common::is_running_in_ci() {
            eprintln!("Skipping external-service test in CI");
            return;
        }

        let manager = common::get_test_manager();

        let _urls = manager
            .start_web_service()
            .await
            .expect("start_web_service failed");

        let catalog = manager.catalog();
        let loaded = catalog
            .get_loaded_models()
            .await
            .expect("get_loaded_models via external service failed");

        // Vec<String> is always a valid list; just ensure the call succeeded.
        let _ = loaded;
    }
}
